File: Baekjoon/Class3/19.1260 DFS와 BFS.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split(/\s+/).map(Number);

const out = []

let cursor = 0;
const n = input[cursor++];
const m = input[cursor++];
const v = input[cursor++];

graph = Array.from({ length: n + 1 }, () => new Array()); // 무방향성 그래프
for(let _ = 0; _ < m; _++) {
    const start = input[cursor++];
    const end = input[cursor++];
    graph[start].push(end);
    graph[end].push(start);
}
for(const node of graph) // 정렬
    node.sort((a, b) => a - b);

// 깊이 우선 탐색
const dfsRecursive = (start, visited) => {
    visited[start] = true;
    out.push(start);
    for(const end of graph[start])
        if(!visited[end])
            dfsRecursive(end, visited);
}

const dfs = (start) => {
    const visited = new Array(graph.length).fill(false);
    dfsRecursive(start, visited);
}

// 너비 우선 탐색
const bfs = (start) => {
    let qHead = 0;
    const q = [start];
    const visited = new Array(graph.length).fill(false);
    visited[start] = true;
    out.push(start);
    while(qHead < q.length) {
        const start = q[qHead++];
        for(const end of graph[start])
            if(!visited[end]) {
                q.push(end);
                visited[end] = true;
                out.push(end);
            }
    }
}

dfs(v);
console.log(out.join(" "));

out.length = 0;
bfs(v);
console.log(out.join(" "));

File: Baekjoon/Class3/19.1260 DFS와 BFS.py
from collections import deque
import sys
input = sys.stdin.readline
write = sys.stdout.write

n, m, v = map(int, input().split())

graph = [[] for _ in range(n + 1)] #무방향성 그래프
for _ in range(m):
    start, end = map(int, input().split())
    graph[start].append(end)
    graph[end].append(start)
for node in graph: #정렬
    node.sort()

#깊이 우선 탐색
def dfs_recursive(start, visited):
    visited[start] = True
    write(f"{start} ")
    for end in graph[start]:
        if not visited[end]:
            dfs_recursive(end, visited)

def dfs(start):
    visited = [False] * len(graph)
    dfs_recursive(start, visited)
    write("\n")

#너비 우선 탐색
def bfs(start):
    d = deque([start])
    visited = [False] * len(graph)
    visited[start] = True
    write(f"{start} ")
    while d:
        start = d.popleft()
        for end in graph[start]:
            if not visited[end]:
                d.append(end)
                visited[end] = True
                write(f"{end} ")
    write("\n")

dfs(v)
bfs(v)